core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -623,37 +683,114 @@ class LogManager(logDirs: Array[File],
   }
 
   /**
+    * Mark the partition directory in the source log directory for deletion and
+    * rename the future partition in the destination log directory to be the primary partition
+    *
+    * @param topicPartition TopicPartition that needs to be swapped
+    */
+  def deleteAndSwapFutureLog(topicPartition: TopicPartition): Unit = {
```

Comment:
Good point regarding the cleaner checkpoint. I am wondering if it may be better to simply pause log cleaner for all those partitions that are being moved to other log directories. 

This approach may have two advantages over the suggested approach of saving copying the checkpoint file at the beginning of the data movement:

- The broker performance is better than the suggested approach. With the suggested approach, broker needs to clean the data after the checkpoint twice. It is better to only do log cleaning after the partition has been moved to the destination log directory.

- The implementation is simpler than the suggested approach. With the suggested approach, the offset for the same partition will be recorded twice in two log directories. And we will need to have more logic in LogCleanerManager.allCleanerCheckpoints() to differentiate between checkpoint of primary log and the checkpoint of the future log. And we probably need to take more care to keep the checkpoints across log directories consistent if e.g. log renames during swap phase failed.

The only concern may be that the size of the compacted partition in the source log directory may be larger than if it is compacted. It is only a concern if replica movement takes a long time. And if this is the case, we probably already have problem with the size of the future log in the destination log directory anyway. What do you think?

Reply:
Yes, that's not a bad idea. We could pause log cleaning during disk movement and copy the cleaning point after the movement completes.